app/celery/nightly_tasks.py

```diff
@@ -334,3 +335,42 @@ def save_daily_notification_processing_time(bst_date=None):
             messages_within_10_secs=result.messages_within_10_secs,
         )
     )
+
+
+@notify_celery.task(name="delete-oldest-quarter-of-unneeded-notification-history")
+def delete_oldest_quarter_of_unneeded_notification_history():
+    # This retention limit is hardcoded for the moment. It is picked from
+    # https://github.com/alphagov/notifications-aws/blob/main/decisions/2022-12-01-notification-history-retention-period.md
+    # and can be increased in the future when we move 3 months into the next financial year
+    retention_limit = datetime(2022, 4, 1)
+
+    oldest_record = db.session.query(func.min(NotificationHistory.created_at)).one()[0]
+    if oldest_record >= retention_limit:
+        current_app.logger.info(
+            "No notifications older than retention date of %s so nothing to delete", retention_limit
+        )
+        return
+
+    # Pick a potential deletion date to start, that we know is older than the oldest notification
+    # in notification_history at the time of writing this code
+    deletion_date = datetime(2019, 10, 1)
+    while deletion_date < oldest_record:
+        # There are no notifications older than our deletion date, lets add 3 months
+        # and see if that is still true
+        # When we finally find a deletion date that has notifications older than the deletion date
+        # we can then delete anything older (ie the previous quarters data)
+        deletion_date = deletion_date + dateutil.relativedelta.relativedelta(months=3)
+
+    delete_notification_history_older_than_datetime(deletion_date)
+
+
+def delete_notification_history_older_than_datetime(older_than_datetime):
+    current_app.logger.info("Beginning to delete notification_history older than %s", older_than_datetime)
+
+    num_rows_deleted = NotificationHistory.query.filter(
+        NotificationHistory.sent_at < older_than_datetime,
```

Comment:
I kind of wasn't sure which to use out of the three of them. I think I very briefly thought, at which point does a notification become 'billable' and I guessed that would be when it is sent.

It definitely does make sense that it should be consistent with the value we pick in `oldest_record`. 

Giving it some more thought, I think I should make it `created_at`.

For our code that populates the `ft_billing` table, we decide which notifications are eligible for billing for a particular day based on their `created_at` - https://github.com/alphagov/notifications-api/blob/main/app/dao/fact_billing_dao.py#L552

Letters does appear to be the exception as you mention, but I think we should go for consistency with our ft_billing table because we are picking our retention date based on our billing needs (ie the ft billing table).

I'll address what we do about letters in a separate comment... 

As an aside, it looks like we still write into ft_billing for letters based on the Notification.created_at date (https://github.com/alphagov/notifications-api/blob/main/app/dao/fact_billing_dao.py#L618) but now are also writing into FactBillingLetterDespatch table based on the dispatch date for letters. I'm a bit confused about why we are still writing into ft_billing for letters if we have this new FactBillingLetterDespatch table which I'm guessing we are now using for billing. Feels like we are left with two sources of data which might get confused? Any idea if we are going to reduce these down to a single one in time?

Reply:
I've added a commit which changes it to use `created_at` rather than `sent_at`

Reply:
I agree that `created_at` sounds good, and letters can be considered separately. I haven't heard of any plans to not have letters included in both ft_billing and ft_billing_letter_despatch

Reply:
> I'm a bit confused about why we are still writing into ft_billing for letters if we have this new FactBillingLetterDespatch table which I'm guessing we are now using for billing. Feels like we are left with two sources of data which might get confused? Any idea if we are going to reduce these down to a single one in time?

I think it's because we use `FactBillingLetterDespatch` for us being billed by DVLA, and `ft_billing` for us billing the service users. It's because we bill our users the same way we used to, but the way we get billed by DVLA has changed.